### ISSUE TITLE
Feature Request: Reduce "clutter" in lock file

### ISSUE THREAD

**Issue by njt1982** [composer/composer#11378]
This may have already been discussed but I couldn't find anything relevant, so apologies if it has.

TLDR: I propose that lock file should only contain exactly what it *needs* to do its job; lock a package to its required version.

At minimum this should be the name and version.

I suspect some others, like source and/or dist might be too (to avoid composer having to look it up when installing?) and maybe the type or autoload info so install can make the autoload.php and/or copy files to the right place.

The main things I'd like to see "gone" are essentially all "meta data". For example:

- Authors - lovely info; its always nice to credit people's work! 🙂But why is it needed in the lock file?
- License - Maybe this is a legal thing?
- Description, homepage, keywords, support, notification-uri.... all these look like "metadata" to me and are surely not needed as part of the lock file? I get why they'd be in the source composer.json as I suspect things like this are needed when maintaining the repository?

If you look a:
- Gemfile.lock - it's just a nested list of package and version, nothing else. Bundler can infer everything else during install / update.
- `package-lock.json` - npm stored a minimal block per package; name, version, hash for integrity per file, resolved URL to download.. occasionally it puts in some requires/dependencies too (although I'd argue this is something only needed during an update (or addition) in which case it could be recalculated at that point? But maybe this is an optimisation)
- `yarn.lock` - basically the same as npm.
- `pyproject.toml` - (eg used by Poetry) or even pip lockfiles... VERY minimal list, similar to Gemfile.

I can understand some data being in the lockfile for optimisations (load download URL) and to aid in building the autoloader quickly... but I think it would be beneficial to drop a lot of the metadata.

Reason?
I frequently see PR's for projects I work on where a lock file is updated and gets "littered" (for want of a better word) with insignificant changes. When installing a package, does it matter if the author has changed if the package itself has not? They also bloat the lockfile to be unnecessarily large (again, making diff'ing trickier than it needs to be).
For example, today a colleague used the patches plugin for something and did a `composer update --lock` to refresh the hash... this then went through and added a new author to many projects.

Maybe I'm missing something here, so happy to be corrected :)

**Comment by njt1982** [comment]
> without loading metadata from packagist again

What would it need to load?

npm _just_ (simplification time!) stores a URL to download.

Is this to do with things like autoloader generation, as I understand composer does a bit more than some other package manager (like npm!)...

As I said, I understand that some of the stuff would need to be kept. I suspect a simple `name => version` map would be insufficient here.

But does it need homepage URL, authors and things? 🤷🏻‍♂️

All the "excess" simply adds another chance for the lock file to have changes which aren't needed and cause "noise" in PRs?

**Comment by stof** [comment]
`composer install` verifies compatible dependencies and then generates some extra things (the platform checks and the autoloader for instance). And given that composer allows partial updates, it also needs to have *all* metadata impacting the updates (in case the locked package is kept locked in a partial update).

Among the existing information, the things that could probably be removed would be `require-dev` (as the dev requirements of dependencies are never used).
`authors` and `support` could be removed at the expense of not showing them anymore when you run `composer show some/package`
Other info are used by composer (for instance, the homepage is used when your terminal supports links to make `composer outdated` and a few other command link to the package when showing the package name).

**Comment by njt1982** [comment]
> when you run composer show some/package
> the homepage is used when your terminal supports links to make composer outdated

These all feel like things that could be looked up when you run that thing as a one-off? I mean it's an assumption that the authors, support and URL are still accurate at the time of running that command if you just regurgitate whatever is in the lockfile which was generated last time it was updating that package?

> composer install verifies compatible dependencies 

Surely this is a job that should happen once during an update/require? (I'm not familiar with the internal workings of composer, so forgive me if I'm wrong/oversimplifying things in my head!)

I appreciate that removing _all_ metadata might be unfeasible (at least at this stage!)... I just personally find it unnecessary noise to store things like author and homepage in there... then, in the event that a minor lock file update happens I end up with loads of unrelated PR changes because Claire has taken over from Joe on the project 😉

**Comment by stof** [comment]
A `composer install` still runs the solver (but with only the locked packages as input).

Note that for a given package, authors should not change in the lock file if the version is not updated AFAICT.

**Comment by Seldaek** [comment]
IMO the trade-off here isn't worth it. The lock file isn't really meant to be reviewed as a text file, and in npm changes usually you end up with so many packages changing that it's pretty hard to review as well anyway..

You'd gain a bit less diffs but reduce the UX in other aspects as discussed above, sure it's fixable for most cases with added network latency to sync up, which is also not great UX..

A little plug here, if you use Private Packagist we do offer a way more readable diff (with lots of additional data) for lock file changes right in your PRs https://packagist.com/docs/update-review

**Comment by njt1982** [comment]
> The lock file isn't really meant to be reviewed as a text file

And yet, still, it appears as a Diff in a PR ;) 

Also, given that human review pull requests, doesn't it make sense that this file be easily parsable by a human to confirm that the changes being suggested make sense? Currently with all the "noise" it's not trivial to spot if something has been accidentally downgraded (etc) amongst all the other "irrelevant" updates caused by license and author syncing. I often find composer.lock file changes a case of "cant see the wood for the trees".

**Comment by stof** [comment]
@njt1982 there are tools to give a human-readable summary of a lock diff (showing you which packages were updated):

- Private Packagist has a GitHub integration showing it in a comment (see https://github.com/composer/packagist/pull/1355#issuecomment-1386873973 for an example)
- https://github.com/davidrjonas/composer-lock-diff is a CLI tool to do such comparison as well when you don't have Private Packagist (it supports outputting in a markdown format to copy/paste it in a GitHub comment as well)
